interface/src/app/signature/enriched_signatures.js

```diff
@@ -0,0 +1,203 @@
+
+    // Do nothing if no genes are specified in URL.
+    if (!$stateParams.genes || !$stateParams.genes.split(',').length) {
+      self.statusMessage = 'No genes are specified.';
+      self.enrichedSignatures = [];
+      return;
+    }
+
+    var genesInUrl = [];
+    $stateParams.genes.split(',').forEach(function(token) {
+      var id = parseInt(token);
+      if (!isNaN(id) && genesInUrl.indexOf(id) === -1) {
+        genesInUrl.push(id);
+      }
+    });
+
+    // Promise that gets all signatures of given mlmodel in URL:
+    var p1 = Signature.get(
+      {mlmodel: self.modelInUrl, limit: 0},
+      function success(response) {
+        self.signatures = Object.create(null);
+        response.objects.forEach(function(element) {
+          self.signatures[element.id] = element.name;
+        });
+      },
+      function error(err) {
+        var message = errGen('Failed to get signatures: ', err);
+        $log.error(message);
+        self.statusMessage = message + '. Please try again later.';
+      }
+    ).$promise;
+
+    // Promise that gets the total number of genes:
+    var p2 = Gene.get(
+      {},
+      function success(response) {
+        self.geneNum = response.meta.total_count;
+      },
+      function error(err) {
+        var message = errGen('Failed to get total gene number: ', err);
+        $log.error(message);
+        self.statusMessage = message + '. Please try again later.';
+      }
+    ).$promise;
+
+    var participations;
+    // Promise that gets all participation records that are related to
+    // the genes in URL:
+    var p3 = Participation.get(
+      {'related_genes': $stateParams.genes, 'limit': 0},
+      function success(response) {
+        participations = response.objects;
+      },
+      function error(err) {
+        var message = errGen(
+          'Failed to get signature-gene participations: ', err);
+        $log.error(message);
+        self.statusMessage = message + '. Please try again later.';
+      }
+    ).$promise;
+
+    var pValueSigDigits = 3;
+    // groupedGenes is an object that behaves like a 2-D table of genes.
+    // groupedGenes[participationTypeID][signatureID] is an object whose
+    // keys are gene IDs and values are genes names of those genes that
+    // participate with signatureID and with participationTypeID.
+    var groupedGenes = Object.create(null);
+
+    // Helper function that calculates the enrichment for each signature
+    // that includes at least one gene in the url.
+    var calculateEnrichments = function(typeName, genesBySignatures) {
+      var m = genesInUrl.length;
+      var N = self.geneNum;
+      var pValueArray = [];
+      var matchedGenesBySignature = [];
+      var signatureIDs = Object.keys(genesBySignatures);
+      signatureIDs.forEach(function(sigID) {
+        var genes = genesBySignatures[sigID];
+        var n = Object.keys(genes).length, k = 0;
+        var matchedGenes = [], selectedGene;
+        var pValue;
```

Comment:
I know this is a lot of `var`s to define, but could we please have just one per line? That makes it easier to scan and find where each one is defined. If you don't like writing `var` on each line, you could say:

```javascript
var genes = genesBySignatures[sigID],
  n = Object.keys(genes).length,
  k = 0,
  matchedGenes = [],
  selectedGene,
  pValue;
```

Reply:
Sure!